policy-controller/k8s/index/src/index.rs

```diff
@@ -0,0 +1,738 @@
+#[derive(Debug)]
+struct Pod {
+    meta: pod::Meta,
+
+    /// The pod's named container ports. Used by `Server` port selectors.
+    ///
+    /// A pod may have multiple ports with the same name. E.g., each container
+    /// may have its own `admin-http` port.
+    port_names: HashMap<String, pod::PortSet>,
+
+    /// All known TCP server ports. This may be updated by
+    /// `Namespace::reindex`--when a port is selected by a `Server`--or by
+    /// `Namespace::get_pod_server` when a client discovers a port that has no
+    /// configured server (and i.e. uses the default policy).
+    port_servers: pod::PortMap<PodPortServer>,
+}
+
+/// Holds the state of a single port on a pod.
+#[derive(Debug)]
+struct PodPortServer {
+    /// The name of the server resource that matches this port. Unset when no
+    /// server resources match this pod/port (and, i.e., the default policy is
+    /// used).
+    name: Option<String>,
+
+    /// A sender used to broadcast pod port server updates.
+    tx: watch::Sender<InboundServer>,
+
+    /// A receiver that is updated when the pod's server is updated.
+    rx: watch::Receiver<InboundServer>,
+}
+
+/// Holds the state of policy resources for a single namespace.
+#[derive(Debug)]
+struct PolicyIndex {
+    cluster_info: Arc<ClusterInfo>,
+    servers: HashMap<String, server::Server>,
+    server_authorizations: HashMap<String, server_authorization::ServerAuthz>,
+}
+
+// === impl Index ===
+
+impl Index {
+    pub fn shared(cluster_info: impl Into<Arc<ClusterInfo>>) -> SharedIndex {
+        let cluster_info = cluster_info.into();
+        Arc::new(RwLock::new(Self {
+            cluster_info: cluster_info.clone(),
+            namespaces: NamespaceIndex {
+                cluster_info,
+                by_ns: HashMap::new(),
+            },
+        }))
+    }
+
+    /// Obtains a pod:port's server receiver.
+    ///
+    /// An error is returned if the pod is not found. If the port is not found,
+    /// a default is server is created.
+    pub fn pod_server_rx(
+        &mut self,
+        namespace: &str,
+        pod: &str,
+        port: u16,
+    ) -> Result<watch::Receiver<InboundServer>> {
+        let ns = self
+            .namespaces
+            .by_ns
+            .get_mut(namespace)
+            .ok_or_else(|| anyhow::anyhow!("namespace not found: {}", namespace))?;
+        let pod = ns
+            .pods
+            .by_name
+            .get_mut(pod)
+            .ok_or_else(|| anyhow::anyhow!("pod {}.{} not found", pod, namespace))?;
+        Ok(pod
+            .port_server_or_default(port, &self.cluster_info)
+            .rx
+            .clone())
+    }
+}
+
+impl kubert::index::IndexNamespacedResource<k8s::Pod> for Index {
+    #[instrument(skip_all, name = "pod.apply", fields(
+        ns = %pod.namespace().unwrap(),
+        pod = %pod.name(),
+    ))]
+    fn apply(&mut self, pod: k8s::Pod) {
+        let namespace = pod.namespace().unwrap();
+        let name = pod.name();
```

Comment:
nit/TIOLI: since these are synchronous functions, it would also be fine to use the `Span::entered` method, so we could use the function-like span macros here. that might be a bit less verbose than using `#[instrument]` in this case:
```suggestion
    fn apply(&mut self, pod: k8s::Pod) {
        let namespace = pod.namespace().unwrap();
        let name = pod.name();
        let _span =  tracing::trace_span!("pod.apply", ns = %namespace, pod = %name).entered();
```